heron/proto/ckptmgr.proto

```diff
@@ -0,0 +1,200 @@
+message CleanStatefulCheckpointResponse {
+  required heron.proto.system.Status status = 1;
+  repeated string cleaned_checkpoints = 2;
+}
+
+// stmgr -> ckptmgr messages
+
+// This is the message that a stmgr sends
+// when it wants to establish the connection
+// with checkpoint manager
+message RegisterStMgrRequest{
+  required string topology_name = 1;
+  required string topology_id = 2;
+  required string stmgr_id = 3;
+}
+
+// This is the message that checkpoint manager
+// sends when it receives the register request
+// from stmgr
+message RegisterStMgrResponse {
+  required heron.proto.system.Status status = 1;
+}
+
+// This is the request that StMgr sends to Checkpoint Mgr
+// to store instance checkpoint data
+message SaveInstanceStateRequest {
+  // Information about the instance whose state this is
+  required heron.proto.system.Instance instance = 1;
+  required InstanceStateCheckpoint checkpoint = 2;
+}
+
+// After successfully saving state, this is what
+// Checkpoint Mgr sends back as response
+message SaveInstanceStateResponse {
+  required heron.proto.system.Status status = 1;
+  required heron.proto.system.Instance instance = 2;
+  required string checkpoint_id = 3;
+}
+
+// This is the request that StMgr sends to Checkpoint Mgr
+// to fetch the instance state data
+message GetInstanceStateRequest {
+  required heron.proto.system.Instance instance = 1;
+  required string checkpoint_id = 2;
+}
+
+// This is the response that Checkpoint Mgr sends to StMgr
+// for its get instance state request
+message GetInstanceStateResponse {
+  required heron.proto.system.Status status = 1;
+  required heron.proto.system.Instance instance = 2;
+  required string checkpoint_id = 3;
+  optional InstanceStateCheckpoint checkpoint = 4;
+}
+
+// stmgr -> Instance messages
+
+// This is the message that the stmgr sends to its
+// local tasks to begin initiating checkpoint
+message InitiateStatefulCheckpoint {
+  required string checkpoint_id = 1;
+}
+
+// This is the message that the instance sends
+// when it wants stmgr to store its state
+message InstanceStateCheckpoint {
+  required string checkpoint_id = 1;
+  required bytes state = 2;
+}
+
+// This is the message that stmgr sends to its instace
+// asking them to restore their state
+message RestoreInstanceStateRequest {
+  required InstanceStateCheckpoint state = 1;
+}
+
+// This is the message that instance sends to its stmgr
+// when done with restoring the state
+message RestoreInstanceStateResponse {
+  required heron.proto.system.Status status = 1;
+  required string checkpoint_id = 2;
+}
+
+// Message sent by stmgr to Instance to start processing
+// This is related to StartStmgrStatefulProcessing message above
+// but between stmgr -> instance.
+message StartInstanceStatefulProcessing {
+  required string checkpoint_id = 1;
+}
+
+// stmgr -> stmgr messages
+
+// Once an instance is done checkpointing, this message is sent
```

Comment:
Couple of clarifications:-
1. Just to be clear in Heron parlance component means one logical component(like spout/bolt). Instance is one running process of it. Thus there are 5 instances of 'wordSpout' component. In this case, it is instance level.
2. Chandy Lamport's algorithm necessitates that downstream instances are aware of all possible upstreamers. In practice this is performed at the stmgr so the real instance code is not aware of it. We cannot involve tmaster here because then a) it is not decentralized and b) only local state inside stmgr is sufficient